Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+//
+//  HTTPServerPipelineHandler.swift
+//  NIOHTTP1
+//
+//  Created by Cory Benfield on 01/03/2018.
+//
+
+import NIO
+
+/// A `ChannelHandler` that handles HTTP pipelining by buffering inbound data until a
+/// response has been sent.
+///
+/// This handler ensures that HTTP server pipelines only process one request at a time.
+/// This is the safest way for pipelining-unaware code to operate, as it ensures that
+/// mutation of any shared server state is not parallelised, and that responses are always
+/// sent for each request in turn. In almost all cases this is the behaviour that a
+/// pipeline will want.
+public class HTTPServerPipelineHandler: ChannelInboundHandler, ChannelOutboundHandler {
+    public typealias InboundIn = HTTPServerRequestPart
+    public typealias InboundOut = HTTPServerRequestPart
+    public typealias OutboundIn = HTTPServerResponsePart
+    public typealias OutboundOut = HTTPServerResponsePart
+
+    public init() { }
+
+    /// Whether the channel pipeline is currently processing a response.
+    private var responsePending = false
+
+    /// While we're waiting to send the response we don't read from the socket.
+    /// This keeps track of whether we need to call read() when we've send our response.
+    private var readPending = false
+
+    /// The buffered HTTP requests that are not going to be addressed yet.
+    private var requestBuffer: [NIOAny] = []
```

Comment:
I wonder if we want to use a `CircularBuffer` here ?. This sounds like a better fit in terms of overhead. WDYT ?